src/plugins/data/public/search/search_service.ts

```diff
@@ -73,63 +42,45 @@ export class SearchService implements Plugin<ISearchSetup, ISearchStart> {
+  private registerSearchStrategyProvider = <T extends TStrategyTypes>(
+    name: T,
+    strategyProvider: TSearchStrategyProvider<T>
+  ) => {
+    this.searchStrategies[name] = strategyProvider;
+  };
 
-  public setup(core: CoreSetup, packageInfo: PackageInfo): ISearchSetup {
-    const search = (this.search = createAppMountSearchContext(this.searchStrategies).search);
-    core.application.registerMountContext<'search'>('search', () => {
-      return { search };
-    });
+  private getSearchStrategy = <T extends TStrategyTypes>(name: T): TSearchStrategyProvider<T> => {
+    const strategyProvider = this.searchStrategies[name];
+    if (!strategyProvider) {
+      throw new Error(`Strategy with name ${name} does not exist`);
+    }
+    return strategyProvider;
+  };
 
-    this.contextContainer = core.context.createContextContainer();
+  public setup(core: CoreSetup, packageInfo: PackageInfo): ISearchSetup {
     this.esClient = getEsClient(core.injectedMetadata, core.http, packageInfo);
 
-    const registerSearchStrategyProvider: TRegisterSearchStrategyProvider = <
-      T extends TStrategyTypes
-    >(
-      plugin: PluginOpaqueId,
-      name: T,
-      strategyProvider: TSearchStrategyProvider<T>
-    ) => {
-      this.searchStrategies[name] = this.contextContainer!.createHandler(plugin, strategyProvider);
-    };
+    this.registerSearchStrategyProvider(SYNC_SEARCH_STRATEGY, syncSearchStrategyProvider);
 
-    const api = {
-      registerSearchStrategyContext: this.contextContainer!.registerContext,
-      registerSearchStrategyProvider,
-      __LEGACY: {
-        esClient: this.esClient,
-      },
-    };
+    this.registerSearchStrategyProvider(ES_SEARCH_STRATEGY, esSearchStrategyProvider);
 
-    api.registerSearchStrategyContext(this.initializerContext.opaqueId, 'core', () => core);
-    api.registerSearchStrategyProvider(
-      this.initializerContext.opaqueId,
-      SYNC_SEARCH_STRATEGY,
-      syncSearchStrategyProvider
-    );
-
-    // ES search capabilities are written in a way that it could easily be a separate plugin,
-    // however these two plugins are tightly coupled due to the default search strategy using
-    // es search types.
-    esSearchService(this.initializerContext).setup(core, { search: api });
-
-    return api;
+    return {
+      registerSearchStrategyProvider: this.registerSearchStrategyProvider,
+    };
   }
 
   public start(core: CoreStart): ISearchStart {
-    if (!this.search) {
-      throw new Error('Search should always be defined');
-    }
     return {
-      search: this.search,
+      search: (request, options, strategyName) => {
+        const strategyProvider = this.getSearchStrategy(strategyName || DEFAULT_SEARCH_STRATEGY);
+        const strategyPromise = strategyProvider({
+          core,
+          getSearchStrategy: this.getSearchStrategy,
+        });
+        return strategyPromise.search(request as any, options);
```

Comment:
I really think we can eventually massage the types to get rid of this `any` but I think we'll have to adjust some things so I'm happy to leave it out of the scope of this PR.